text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,425 @@
+
+As already stated, the current `Range` implementation treats the occurrence of `+-Inf` or `NaN` parameters in *any* of the `min, max, step` parameters as sufficient condition for *infinite*. This RFC considers the opposite: the no-progress expression cannot be meaningfully evaluated if any one of the 3 tested parameters is a `NaN` value, and one can also not decide whether any .next() iterations which add NaN lie within the given numerical range, or, if `min` or `max` are `NaN`, what that range is in the first place. Any occurrence of `NaN` in the Range parameters therefore produces an *empty* Range. Similarly, while a `step` argument of `+-Inf` may "pass" the inspection by the *no-progress criterion*, no finite points within the mathematical range `[min, max)` can be computed, including when `min, max` themselves are `inf`. Therefore, such Ranges are *empty*, too. A `step` argument of `0` satisfies the *no-progress criterion* and produces an *empty* Range.
+
+The cases that would remain *infinite* under this proposal are then only those which fulfill the following condition: they are *not empty* (as per the evaluation above) AND their upper bound is either `+Inf` or `-Inf`. Typical examples would be `Range[F64](0, inf, 1)` or `Range[F64](0, -inf, -1)`. We can't allow a lower bound that is not finite, because then the elements returned on .next() calls would again not be finite.
+
+Summarizing this, the proposed algorithm for the Range classification is outlined in a close-to-code form. The following expressions are useful:
+
+```
+no_progress = ((min <= max) and (step <= 0) or (min >= max) and (step >= 0))
+progress = not no_progress = ((min > max) or (step > 0)) and ((min < max) or (step < 0))
+                           = (((min > max ) or (step > 0)) and (min < max)) or (((min > max ) or (step > 0)) and (step < 0))
+                           = ((step > 0) and (min < max)) or ((min > max ) and (step < 0))
+                           = ((min < max) and (step > 0)) or ((min > max ) and (step < 0))
+```
+
+The above discussed combined criterion for *empty* was: either the no_progress criterion is true, or the generated values would not be finite. Finite values require both `min` and `step` to be finite, since the values are generated by an iterative addition of `step` to `min`. Therefore, one can state:
+
+```
+not_empty = progress and min.finite() and step.finite() // .finite() => the value is not `+-Inf` and not `NaN`
+infinite = not_empty and max.infinite()
+empty = not not_empty
```

Comment:
I feel like these three distinctions are what your entire RFC is about and yet once I got here I did not feel like you made a strong case for why we should have these three distinctions.

Reply:
These three distinctions are not what the RFC is about -- they are merely used to explain how all that is explained above is implemented. I did not mean to and might indeed not have made a strong case for how I would implement this, although I think it is logical.

Reply:
> This expectation can be met by introducing and in such and similar cases returning *empty* ranges. One unexpected side effect of Pony's current behavior is that even though finite bounds are given, an *infinite* iterator can be created from them which, when used without consideration for this side effect, can produce indefinitely-running actors and "hanging" programs.

The definitions I commented on here seem like the representation of your central point in the introduction. That is, we are carving out three distinct categories of Range to be finite, infinite, or empty. Although, my comment here was made under the "why empty, why not parameter rearrangment" line of thinking which I have now convince myself does not fit with the purpose of Range anyhow. Had I not been in that line of thinking, I would not have made the comment I did here.

I would suggest adding language to the introduction that we are making changes to prevent unsuspecting infinite ranges when, according to construction, progress cannot be made and therefore the range should be empty, not infinite. Then add in alternatives, that preventing such unsuspecting infinite ranges could be done via parameter rearrangement, but due to the half-closed interval that defines Range, such rearrangement is likely to introduce more errors, not fewer. Keeping Range to always move from `min` to `max` and having it be empty when progress cannot be made is a clearer design.